internal/k8s/controller.go

```diff
@@ -1551,36 +1567,41 @@ func (lbc *LoadBalancerController) createVirtualServer(virtualServer *conf_v1alp
 		virtualServerRoutes = append(virtualServerRoutes, vsr)
 
 		for _, u := range vsr.Spec.Upstreams {
-			endpointsKey := configs.GenerateEndpointsKey(vsr.Namespace, u.Service, u.Port)
-			endpoints[endpointsKey] = lbc.getEndpointsForService(vsr.Namespace, u.Service, int(u.Port))
+			svc, err := lbc.getServiceForUpstream(u, vsr.Namespace)
+			if err != nil {
+				glog.V(3).Infof("Error getting service %v: %v", u.Service, err)
+			} else {
+				endpointsKey := configs.GenerateEndpointsKey(vsr.Namespace, u.Service, u.Port)
+				endps, external, err := lbc.getEndpointsForService(vsr.Namespace, u.Service, u.Port, svc)
+
+				if err == nil && external && lbc.isNginxPlus {
+					externalNameSvcs[u.Service] = true
+				}
+				endpoints[endpointsKey] = endps
+			}
 		}
 	}
 
 	virtualServerEx.Endpoints = endpoints
 	virtualServerEx.VirtualServerRoutes = virtualServerRoutes
+	virtualServerEx.ExternalNameSvcs = externalNameSvcs
 
 	return &virtualServerEx, virtualServerRouteErrors
 }
 
-func (lbc *LoadBalancerController) getEndpointsForService(namespace string, name string, port int) []string {
+func (lbc *LoadBalancerController) getEndpointsForService(namespace string, name string, port uint16, svc *api_v1.Service) (result []string, isExternal bool, err error) {
```

Comment:
I think if you call `lbc.getServiceForUpstream()` from this function, rather from than `createVirtualServer`, it will simplify the logic. For example, `getEndpointsForService` will only need `(namespace string, name string, port uint16)` as arguments. Let me know what you think.

Reply:
I'll do it, but then the meaning will change, cos we need the upstream to get the service.